File: source/materials/diffuse_light.cpp
#include "materials/diffuse_light.hpp"
#include "math/details.hpp"
#include "math/ortho_nomal_basis.hpp"

diffuse_light::diffuse_light(const colour &emission_in) noexcept : emission(emission_in) {}

auto diffuse_light::scatter(const ray &ray_in, const hit &hit_in, double u, double v) const
   -> scatter_data
{
   const auto basis = ortho_normal_basis::from_z(hit_in.normal);
   const vec scatter_dir = hemisphere_sample(basis, u, v);

   // clang-format off
   return scatter_data
      {
         .emission = emission,
         .diffuse = {0.0, 0.0, 0.0},
         .scattered_ray = ray(hit_in.position, scatter_dir)
      };
   // clang-format on
}

File: go/materials/diffuse_light.go
package materials

import (
	"go_pt/maths"
	"go_pt/tracer"
)

type DiffuseLight struct {
   emissionColour maths.Vec3
}

func NewDiffuseLightMaterial(emissionColour *maths.Vec3) DiffuseLight {
   return DiffuseLight{emissionColour: *emissionColour}
}

func (mat *DiffuseLight) Scatter(ray *tracer.Ray, hit *tracer.Hit, u, v float64) MaterialScatterData {
   basis := maths.OrthoNormalBasisFromZ(&hit.Normal)
   scatterDir := maths.HemisphereSample(&basis, u, v)

   return MaterialScatterData{
      Emission: mat.emissionColour, 
      Diffuse: maths.Vec3{}, 
      ScatteredRay: tracer.Ray{Origin: hit.Position, Direction: scatterDir}}
}
